src/System.Private.CoreLib/shared/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -0,0 +1,1117 @@
+                    break;
+                keys[lo + i - 1] = keys[lo + child - 1];
+                i = child;
+            }
+            keys[lo + i - 1] = d;
+        }
+
+        private static void InsertionSort(T[] keys, int lo, int hi)
+        {
+            Debug.Assert(keys != null);
+            Debug.Assert(lo >= 0);
+            Debug.Assert(hi >= lo);
+            Debug.Assert(hi <= keys.Length);
+
+            int i, j;
+            T t;
+            for (i = lo; i < hi; i++)
+            {
+                j = i;
+                t = keys[i + 1];
+                while (j >= lo && (t == null || t.CompareTo(keys[j]) < 0))
+                {
+                    keys[j + 1] = keys[j];
+                    j--;
+                }
+                keys[j + 1] = t;
+            }
+        }
+    }
+
+    #endregion
+
+    #region ArraySortHelper for paired key and value arrays
+
+    internal partial class ArraySortHelper<TKey, TValue>
+    {
+        public static void Sort(TKey[] keys, TValue[] values, int index, int length, IComparer<TKey> comparer)
+        {
+            Debug.Assert(keys != null, "Check the arguments in the caller!");  // Precondition on interface method
+            Debug.Assert(values != null, "Check the arguments in the caller!");
+            Debug.Assert(index >= 0 && length >= 0 && (keys.Length - index >= length), "Check the arguments in the caller!");
+
+            // Add a try block here to detect IComparers (or their
+            // underlying IComparables, etc) that are bogus.
+            try
+            {
+                if (comparer == null || comparer == Comparer<TKey>.Default)
+                {
+                    comparer = Comparer<TKey>.Default;
+                }
+
+                IntrospectiveSort(keys, values, index, length, comparer);
+            }
+            catch (IndexOutOfRangeException)
+            {
+                IntrospectiveSortUtilities.ThrowOrIgnoreBadComparer(comparer);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_IComparerFailed, e);
+            }
+        }
+
+        private static void SwapIfGreaterWithItems(TKey[] keys, TValue[] values, IComparer<TKey> comparer, int a, int b)
+        {
+            Debug.Assert(keys != null);
+            Debug.Assert(values != null && values.Length >= keys.Length);
+            Debug.Assert(comparer != null);
+            Debug.Assert(0 <= a && a < keys.Length);
+            Debug.Assert(0 <= b && b < keys.Length);
+
+            if (a != b)
+            {
+                if (comparer.Compare(keys[a], keys[b]) > 0)
+                {
+                    TKey key = keys[a];
+                    keys[a] = keys[b];
+                    keys[b] = key;
+
+                    TValue value = values[a];
+                    values[a] = values[b];
+                    values[b] = value;                    
+                }
```

Comment:
Please refer to corert [here](https://github.com/dotnet/corert/blob/256481c2afd7290600ec4af2b7ac247436233254/src/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs#L394-L399): 
```
                    if (values != null)
                    {
                        TValue value = values[a];
                        values[a] = values[b];
                        values[b] = value;
                    }
```

Reply:
After taking closer look, I think these null checks are unnecessary (both in CoreCLR and CoreRT). It is fine to remove them.

Reply:
@jkotas how about the Debug.Assert statements explained [here](https://github.com/dotnet/coreclr/pull/18125#issuecomment-392153464)?

Reply:
If I am reading the code correctly, I think that the CoreCLR version of these asserts is right. I do not think that path is ever called with `values == null`.